src/main/java/com/example/bank/bankTransfer/transfer/TransferService.java

```diff
@@ -19,71 +22,96 @@
 @Getter
 @Slf4j
 public class TransferService {
```

Comment:
Zobacz jak ta klasa moglaby byc zaimplementowana i jak powinna byc zaimplementowana. 
Nie jest to implementacja idealna bo nie ma tam strategii ale w sposob w jaki to zrobilem pokazuje ze jej tam nie musi byc : 
`@Service
@RequiredArgsConstructor
@Getter
@Slf4j
public class TransferService {

    private final TransferHistoryService transferHistoryService;
    private final TransferHistoryRepository transferHistoryRepository;
    private final AccountRepository accountRepository;
    private final RateClient rateClient;
    private final TransferValidationUtils transferValidationUtils;

    public void bankTransfer(TransferRequest transferRequest) {
        Account senderAccount = accountRepository.findById(transferRequest.senderAccountId()).orElseThrow();
        Account receiverAccount = accountRepository.findById(transferRequest.receiverAccountId()).orElseThrow();
        BigDecimal amount = transferRequest.amount();
        validateAndProcessTransfer(senderAccount, receiverAccount, amount, transferRequest.title());
    }

    private void validateAndProcessTransfer(Account sender, Account receiver, BigDecimal amount, String title) {
        transferValidationUtils.amountValidation(amount);
        if (isDomesticSameCurrencyTransfer(sender, receiver)) {
            transferValidationUtils.balanceValidation(sender.getBalance(), sender.getCurrency(), amount);
            transfer(sender, receiver, amount, title);
        } else {
            BigDecimal amountInReceiverCurrency = exchangeAmount(sender.getCurrency(), receiver.getCurrency(), amount);
            transferValidationUtils.balanceValidation(sender.getBalance(), sender.getCurrency(), amountInReceiverCurrency);
            transfer(sender, receiver, amount, amountInReceiverCurrency, title);
        }
    }

    private boolean isDomesticSameCurrencyTransfer(Account sender, Account receiver) {
        return sender.getAccountNumber().substring(0, 2).equals(receiver.getAccountNumber().substring(0, 2))
                && sender.getCurrency().equals(receiver.getCurrency());
    }

    private void transfer(Account sender, Account receiver, BigDecimal amount, String title) {
        updateBalancesAndSave(sender, receiver, amount);
        createAndSaveTransferHistories(sender, receiver, amount, title, amount);
        logTransferDetails(amount, receiver.getCurrency(), sender, receiver);
    }

    private void transfer(Account sender, Account receiver, BigDecimal senderAmount, BigDecimal receiverAmount, String title) {
        updateBalancesAndSave(sender, receiver, receiverAmount);
        createAndSaveTransferHistories(sender, receiver, senderAmount, title, receiverAmount);
        logTransferDetails(senderAmount, receiver.getCurrency(), sender, receiver);
    }

    private void updateBalancesAndSave(Account sender, Account receiver, BigDecimal amount) {
        sender.setBalance(sender.getBalance().subtract(amount));
        receiver.setBalance(receiver.getBalance().add(amount));
        accountRepository.saveAll(List.of(sender, receiver));
    }

    private void createAndSaveTransferHistories(Account sender, Account receiver, BigDecimal amount, String title, BigDecimal receiverAmount) {
        TransferHistory senderHistory = transferHistoryService.buildSenderAccountTransferHistory(sender, amount, receiver, title);
        TransferHistory receiverHistory = transferHistoryService.buildReceiverAccountTransferHistory(receiver, receiverAmount, sender, title);
        transferHistoryRepository.saveAll(List.of(senderHistory, receiverHistory));
    }

    private BigDecimal exchangeAmount(Currency senderCurrency, Currency receiverCurrency, BigDecimal amount) {
        BigDecimal exchangeRate = calculateExchangeRate(senderCurrency, receiverCurrency);
        return amount.multiply(exchangeRate).setScale(2, RoundingMode.HALF_EVEN);
    }

    private BigDecimal calculateExchangeRate(Currency senderCurrency, Currency receiverCurrency) {
        BigDecimal senderRate = extractCurrencyExchangeRate(senderCurrency);
        BigDecimal receiverRate = extractCurrencyExchangeRate(receiverCurrency);
        return senderRate.divide(receiverRate, 4, RoundingMode.HALF_EVEN);
    }

    private BigDecimal extractCurrencyExchangeRate(Currency currency) {
        return currency.equals(Currency.PLN) ? BigDecimal.ONE : rateClient.getCurrencyRate(currency).getRates().get(0).getMid();
    }

    private void logTransferDetails(BigDecimal amount, Currency currency, Account sender, Account receiver) {
        log.info("Transfer of {} {} from account {} to account {} completed successfully. Sender balance: {}, Receiver balance: {}",
                amount, currency, sender.getId(), receiver.getId(), sender.getBalance(), receiver.getBalance());
    }
}
`